### ISSUE TITLE
ROMS coupling with WW3 and Nesting

### ISSUE THREAD

**Issue by simion1232006** [jcwarner-usgs/COAWST#285]
Dear there,
I have managed to run ROMS coupling with WW3 which works fine without ROMS internal nesting.
[log3.txt](https://github.com/user-attachments/files/16330632/log3.txt)
see the attached log that everything looks good.
HOwever, when I tried to switch on the ROMS nesting option, the model did not even  start. 
Below is the material I prepared:
1. generate scrip_calypso_nest_static.nc with two roms grids and one wave watch 3 grid
[scrip_coawst_Calypso_nest.txt](https://github.com/user-attachments/files/16330654/scrip_coawst_Calypso_nest.txt)

2. modify the head file
[calypso.txt](https://github.com/user-attachments/files/16330658/calypso.txt)

3. recompile the model
I got the exe

4. modify the input files
[Calypso_ROMS_Grid01_NESTING_3DHD.txt](https://github.com/user-attachments/files/16330687/Calypso_ROMS_Grid01_NESTING_3DHD.txt)
[coupling_calypsoNest.txt](https://github.com/user-attachments/files/16330688/coupling_calypsoNest.txt)

6.  rerun the model

here is the error I got: 
![image](https://github.com/user-attachments/assets/50dac4c8-e805-4e16-b714-965995d689d3)

The model stopped at the very beginning. I could not find an example case with both coupling and nesting. The Sandy case only nests SWAN, not ROMS. Could you please provide some advice? Any help are greatly appreciated..

**Comment by jcwarner-usgs** [comment]
yes.  So we just recently add the ability to use nesting with WW3.   
But you will need to use the same number of WW3 grids are you using for ROMS.
So make another WW3 grid.  Then you will need to go thru a bunch of steps for WW3.  I am in the process to update the documentation but give this a try.
you will need to:
-create a ww3_grid.inp for grid 1
-run WW3/build/bin/ww3_grid  /ww3_strt /ww3_prep for grid 1
-then copy mod_def.ww3 to mod_def.grd1, restart.ww3 to restart.grd1, and wind.ww3 to wind.grd1
the do the same for grid 2
-create a ww3_grid.inp for grid 2
-run WW3/build/bin/ww3_grid  /ww3_strt /ww3_prep for grid 2
then copy mod_def.ww3 to mod_def.grd2, restart.ww3 to restart.grd2, and wind.ww3 to wind.grd2
move all those files (*.grd1 and *.grd2) to the same place as the coawstM

then you need to make a ww3_multi.inp file (i copied one into the zip file)
then in the coupling.in you need to point to that ww3_multi.inp, not the ww3_grid.inp.
see if you can get that to work.

[ww3_multi.zip](https://github.com/user-attachments/files/16333020/ww3_multi.zip)

**Comment by simion1232006** [comment]
> yes. So we just recently add the ability to use nesting with WW3. But you will need to use the same number of WW3 grids are you using for ROMS. So make another WW3 grid. Then you will need to go thru a bunch of steps for WW3. I am in the process to update the documentation but give this a try. you will need to: -create a ww3_grid.inp for grid 1 -run WW3/build/bin/ww3_grid /ww3_strt /ww3_prep for grid 1 -then copy mod_def.ww3 to mod_def.grd1, restart.ww3 to restart.grd1, and wind.ww3 to wind.grd1 the do the same for grid 2 -create a ww3_grid.inp for grid 2 -run WW3/build/bin/ww3_grid /ww3_strt /ww3_prep for grid 2 then copy mod_def.ww3 to mod_def.grd2, restart.ww3 to restart.grd2, and wind.ww3 to wind.grd2 move all those files (*.grd1 and *.grd2) to the same place as the coawstM
> 
> then you need to make a ww3_multi.inp file (i copied one into the zip file) then in the coupling.in you need to point to that ww3_multi.inp, not the ww3_grid.inp. see if you can get that to work.
> 
> [ww3_multi.zip](https://github.com/user-attachments/files/16333020/ww3_multi.zip)

Hi John,
My plan is to do nesting for ROMS only,  not for WW3. Your instruction is mainly about WW3. do you think I can simulate ROMS with nested grid while only couple with one layer of WW3. I know this is not a idea case, but I am doing offshore simulation and refinement of WW3 is not super important. 
I thought this would be quite straightforward as now I can do roms nesting which is working and roms  coupling with WW3 which is also working. I am now trying assemble them together but it does not work as expected. please correct me if there are any misunderstandings.

Kind regards

**Comment by jcwarner-usgs** [comment]
i have coded the roms-ww3 to be for 1 grid on each, or multiple grids on each.
I dont have it working for 2 roms grids and 1 ww3 grid.

**Comment by simion1232006** [comment]
> i have coded the roms-ww3 to be for 1 grid on each, or multiple grids on each. I dont have it working for 2 roms grids and 1 ww3 grid.

I did not know that is the way how the model works. Thanks a lot for the clarification.

**Comment by simion1232006** [comment]
> i have coded the roms-ww3 to be for 1 grid on each, or multiple grids on each. I dont have it working for 2 roms grids and 1 ww3 grid.

Hi John, the coupling case is now working all right. really appreciate your help. It is more complicate than I thought. 
I have one last question, for ROMS coupling with wave model, would you recommend WW3 or swan? For ROMS coupling with SWAN, do  I have more flexibilities in ROMS nesting (different grids for wave and hydro)? I am asking as I may need to do double nesting for ROMS.
[log2gridCouple.txt](https://github.com/user-attachments/files/16343388/log2gridCouple.txt)

**Comment by jcwarner-usgs** [comment]
couple of things:
WW3 is starting at
MWW3 calculating for 2004/09/01 08:00:00 UTC at 10:20:40  status [ 0- 0]
and roms is starting at
     0 2004-09-04 00:00:00.00  4.077162E-02  4.645936E+03  4.645977E+03  9.782262E+13  01
so you either need to start ww3 at 0000 or roms at 0800.

I dont have as much experience with WW3 nesting, this is new to me.  The coupling to SWAN has more capabilities with the BBL and Stokes computations, and nearshore physics are more tested.  if you are further offshore, then WW3 could be as good of a choice.

All my coupling between roms and SWAN-or-WW3 require the same number of roms grids and wave grids. The wave physics scale the same as the ocean.

**Comment by simion1232006** [comment]
> couple of things: WW3 is starting at MWW3 calculating for 2004/09/01 08:00:00 UTC at 10:20:40 status [ 0- 0] and roms is starting at 0 2004-09-04 00:00:00.00 4.077162E-02 4.645936E+03 4.645977E+03 9.782262E+13 01 so you either need to start ww3 at 0000 or roms at 0800.
> 
> I dont have as much experience with WW3 nesting, this is new to me. The coupling to SWAN has more capabilities with the BBL and Stokes computations, and nearshore physics are more tested. if you are further offshore, then WW3 could be as good of a choice.
> 
> All my coupling between roms and SWAN-or-WW3 require the same number of roms grids and wave grids. The wave physics scale the same as the ocean.

Ops, just notice the starting time error. my wind inputs span a longer period than ROMS simulation. I did not notice and setup ww3 simulation time to be the same as wind inputs.
Also really appreicate your clarification. I would go ahead with WW3 and ROMS if all coupling require the same number of meshes. 
I am curious whether the model requires exactly the same meshes or only the number of meshes should be the same?

**Comment by jcwarner-usgs** [comment]
you can run roms on a different grid than swan. 
or run roms on a different grid than ww3.
but if you have more than 1 roms grid, then you need to have the same number of ocean grids and wave grids (like 2 roms , 2 swan).  All the grids can be different sizes.
this is just a limitation in the way i developed the code.  i can only do so much, and we run it with the same number of grids for each model.

**Comment by simion1232006** [comment]
> you can run roms on a different grid than swan. or run roms on a different grid than ww3. but if you have more than 1 roms grid, then you need to have the same number of ocean grids and wave grids (like 2 roms , 2 swan). All the grids can be different sizes. this is just a limitation in the way i developed the code. i can only do so much, and we run it with the same number of grids for each model.

Thanks a lot, John. The model is running OK now. Another thing to mention is if we use the same set of meshes for ROMS and ww3, we will not be able to use wave boundary inputs. This is because the nodes in ww3, onced masked as boundary nodes, will not be used in model simulation thereby have no model outputs. This would create problems when coupling with ROMS .  To solve the issue, we will need to addtional layers of cells in ww3 as boundary nodes. 

Kind regards

**Comment by jcwarner-usgs** [comment]
ok.  I am not sure why you needed to add the extra layers.  we are not doing that, and we are not having that problem.
I think it has to deal with setting the '2' (or whatever it is) in the mapsta.  I think we just set it as 1's.
will need to look at what we do, but we do not need to make an extra rows on ww3 grid.

**Comment by simion1232006** [comment]
> ok. I am not sure why you needed to add the extra layers. we are not doing that, and we are not having that problem. I think it has to deal with setting the '2' (or whatever it is) in the mapsta. I think we just set it as 1's. will need to look at what we do, but we do not need to make an extra rows on ww3 grid.

yes, I have set 2 in mapsta as need it to include wave boundary inputs. 
What I did is first create the mapsta including 2s around the model boundary, and then generate boundary inputs files and use ww3_bounc pre-processor to generate nest.ww3 as boundary condition. Is this the wrong approach?

**Comment by jcwarner-usgs** [comment]
i think i need to do some testing on this. 
let me get back to you.

**Comment by simion1232006** [comment]
> i think i need to do some testing on this. let me get back to you.

Hi John, are you suggesting that these boundary nodes are not required, even if we need wave boundary inputs? I have prepared the wave spectral inputs every 50 nodes around the model boundary and also ran the ww3_bounc to create nest.ww3 for inputs. I would like to know the next step.

**Comment by jcwarner-usgs** [comment]
how about you run the model coupled, and show me the results.  then i can help with what you get.
-j

**Comment by simion1232006** [comment]
> how about you run the model coupled, and show me the results. then i can help with what you get. -j



I think my approach of using addtional layer of boundary nodes (set as 2) working fine. please see the attached the results. 
However, I got issue with the nested model which did not seem to take any information from the coarse model.  
Note that my model uses both coupling and nesting with two set of grids for both ROMS and WW3. 


![Calypso_L1_2016_Hs019](https://github.com/user-attachments/assets/540c2d25-e294-4207-9172-718fc14780cb)
![Calypso_L2_2016_Hs019](https://github.com/user-attachments/assets/eacdbe53-1697-434a-a966-48cc9ca887dd)

**Comment by jcwarner-usgs** [comment]
how did you make the child grid?  did you use my create_ww3_grid_files and create_ww3_nest_bound_points?

**Comment by simion1232006** [comment]
> how did you make the child grid? did you use my create_ww3_grid_files and create_ww3_nest_bound_points?

Yes, I used your script to convert the ROMS child grid to ww3. The coarse grid was the same, but I added extra layers so now they are slightly different. For child grid, I did not have the boundary nodes and everything is masked as within computing domian except the land.

**Comment by jcwarner-usgs** [comment]
what does the log.grd2 show?
did you compile with NESTING on?
did you create that ww3_multi.inp file?

**Comment by simion1232006** [comment]
> what does the log.grd2 show? did you compile with NESTING on? did you create that ww3_multi.inp file?
see the log attached. Yes, I did compile the model with nesting on and I also used the ww3_multi.inp 
[log.grd2.txt](https://github.com/user-attachments/files/16499342/log.grd2.txt)

**Comment by jcwarner-usgs** [comment]
can  u upload the ww3_multi.inp file?

**Comment by simion1232006** [comment]
> can u upload the ww3_multi.inp file?

see the attached file. 
[ww3_multi.inp.txt](https://github.com/user-attachments/files/16504038/ww3_multi.inp.txt)

**Comment by jcwarner-usgs** [comment]
that multi inp file looks ok.
how did you make the child grid?  what is on the perimeter of that mapsta?  i forget if you need a 1 or 2 there.

**Comment by simion1232006** [comment]
> that multi inp file looks ok. how did you make the child grid? what is on the perimeter of that mapsta? i forget if you need a 1 or 2 there.

I got 2 around the perimter of the coarse grid (taking boundary inputs) and 1 for child grid.

**Comment by emanzella** [comment]
john asked me to take a look at your problem. can you try putting a 2 around the perimeter of the child grid? i believe that should tell the child grid to get info from the parent grid

**Comment by emanzella** [comment]
re-reading farther up again-- are you having your parent grid generate boundary output points (i.e. generating a nest.ww3) still? if so this is not necessary/will not work. when using ww3_multi all you need to do is define 2 around the perimeter in your mask/mapsta, and then ww3 will automatically find the points on your parent grid that are active boundary points on the child grid. does this make sense?
the nest.ww3 feature is for when you are not concurrently nesting, but instead just trying to run 1 grid with wave info at the boundary from a larger grid run that you have already performed

**Comment by simion1232006** [comment]
> re-reading farther up again-- are you having your parent grid generate boundary output points (i.e. generating a nest.ww3) still? if so this is not necessary/will not work. when using ww3_multi all you need to do is define 2 around the perimeter in your mask/mapsta, and then ww3 will automatically find the points on your parent grid that are active boundary points on the child grid. does this make sense? the nest.ww3 feature is for when you are not concurrently nesting, but instead just trying to run 1 grid with wave info at the boundary from a larger grid run that you have already performed

I did not generate boundary output points for the child grid. All I did is to define 2 around the parent grid to take wave spectral inputs. I did use ww2_bounc to generate nest.ww3 and rename it to nest.grd1 as boundary condition for the parent grid. I did not define 2 around the child grid which I am going to test now as per  suggestion. I will see how it goes and let you know. Again thanks a lot for help from you and John.

**Comment by simion1232006** [comment]
> re-reading farther up again-- are you having your parent grid generate boundary output points (i.e. generating a nest.ww3) still? if so this is not necessary/will not work. when using ww3_multi all you need to do is define 2 around the perimeter in your mask/mapsta, and then ww3 will automatically find the points on your parent grid that are active boundary points on the child grid. does this make sense? the nest.ww3 feature is for when you are not concurrently nesting, but instead just trying to run 1 grid with wave info at the boundary from a larger grid run that you have already performed

I got a error message when using boundary nodes for child grid:
![image](https://github.com/user-attachments/assets/302647b9-7b2f-4013-8041-6c7fa501d054)

**Comment by emanzella** [comment]
i am not sure, but could be an issue with having open boundary data points on the child that do not exist or are on land mask for the parent. look where that point (1,67) is and see if that tells you anything

**Comment by jcwarner-usgs** [comment]
looks like grid 2, x=1, y=67.  that looks like the first grid point on the child left side that is land.
i got this to work at one point, but i dont remember exactly the issue.
when you created the mapsta, did you use (in my create_ww3_grid_files.m)

ww3_mask=ones(LP,MP);
%
% now make boudaries = 2
%
ww3_mask(:,1)=2;
ww3_mask(:,end)=2;
ww3_mask(1,:)=2;
ww3_mask(end,:)=2;
%
% now remove land and set these points as 0
%
for index1 = 2:MP-1;
    for index2 = 2:LP-1;
        if (mask_rho(index2,index1)==0)
          ww3_mask(index2,index1)=0;
        end
        if (h(index2,index1)<min_depth)
%          ww3_mask(index2,index1)=0;
        end
    end
end
%

i forget if we need to do this or not.  in other words, can you recreate the parent and the child grids with:
&          ww3_mask(index2,index1)=0;

and see if that works?

**Comment by emanzella** [comment]
the current indices john wrote just looks at land masking the interior points and excludes the perimeter boundary points
so you could try doing above but with:

index1= 1:MP
index2=1:LP

I know when you are using a nest.ww3 file you want your whole perimeter to be open boundary points, regardless of whether they are on land. but it might be different when you are concurrently nesting with ww3_multi

**Comment by simion1232006** [comment]
> the current indices john wrote just looks at land masking the interior points and excludes the perimeter boundary points so you could try doing above but with:
> 
> index1= 1:MP index2=1:LP
> 
> I know when you are using a nest.ww3 file you want your whole perimeter to be open boundary points, regardless of whether they are on land. but it might be different when you are concurrently nesting with ww3_multi

I have editted the masking around the child grid, It now has 2 for open boundary and 0 for land masking. Attached is my log of model. I notice that the model considers two group, even I only specified one group in ww3_multi. inp. I am not sure whether this is the reason.

>   Input grid information : 
>   nr extension  lev.   cur.   wind   ice    tau    rho    data
>  ------------------------------------------------------------
>  ------------------------------------------------------------
> 
>   Wave grid information : 
>   nr extension  lev.   cur.   wind   ice    tau    rho    data   move1 rnk grp dmp
>  ---------------------------------------------------------------------------------
>    1 grd1       native native native ---    ---    ---    - - -  no     1   1   -
>    2 grd2       native native native ---    ---    ---    - - -  no     2   2   -
>  ----------------------------------------------------------------------------------
> 
>   Group information : 
>   nr   grids (part of comm.)
>  --------------------------------------------------------------------
>    1    1 (0.0000-1.0000)
>    2    2 (0.0000-1.0000)
>  --------------------------------------------------------------------
> 
> 
> 
>           Initializing model input ...
>             winds        : file OK
>             ice fields   : file not needed
>             momentum     : file not needed
>             air density  : file not needed
> 
>   Computing relation to higher ranked grids
> 
> 
>  WARNING in WMGHGH:   NBI=0 AND RANK > 1 
> 
>   Finished computing relation to higher ranked grids
> 
> 
>   Additional group information
> 
>   No lower rank grid dependencies
> 
>   No same rank grid dependencies
> 
>   No higher rank grid dependencies
> 
>   Running the model :
>  --------------------------------------------------


[log_coupledNested_2016.txt](https://github.com/user-attachments/files/16538376/log_coupledNested_2016.txt)

**Comment by simion1232006** [comment]
> the current indices john wrote just looks at land masking the interior points and excludes the perimeter boundary points so you could try doing above but with:
> 
> index1= 1:MP index2=1:LP
> 
> I know when you are using a nest.ww3 file you want your whole perimeter to be open boundary points, regardless of whether they are on land. but it might be different when you are concurrently nesting with ww3_multi


The solution you suggested is working! I need to mask out the land nodes in child grid where the parent grid is masked as land. The model is now  passing information around the child grid.
But now I observed new errors as attached. Could you please help.
![image](https://github.com/user-attachments/assets/2336093e-8e65-4b25-a573-8ca6ebe4e497)
![image](https://github.com/user-attachments/assets/a3783b4d-a7cf-4658-95dc-3ba36af53c8d)

**Comment by jcwarner-usgs** [comment]
what if you dont mask anything?  no mask for parent and no mask for child.

**Comment by simion1232006** [comment]
> what if you dont mask anything? no mask for parent and no mask for child.

I see the same error even masking everyting as 1 except the perimeter (2). Is there anything I need to modify in the script input?
![image](https://github.com/user-attachments/assets/1b41b519-60ac-41ab-86ac-dfc6ce0b980a)

**Comment by simion1232006** [comment]
> what if you dont mask anything? no mask for parent and no mask for child.
Hi John,
I think I found the issue. I need to shift the grid by very tinny distance as per WW3 manual to avoid the segment errors.The model is now running. I will see how it goes and provide updates.
Also some manual editing of the child grid masking might be required to match the nodes of data transfer between parent and child grids.
![image](https://github.com/user-attachments/assets/7d34ed1a-c53f-4383-84fe-1042633524a6)

**Comment by jcwarner-usgs** [comment]
wow! i did not know that WW3 did that. But we do that.
that was going to be my next ask.
When you create the child grid, in my m file, after you netcdf load the grid file, you can add
lon_rho=lon_rho+0.00001;
and then continue making the child grid.
see if that works, or if that WW3 flag works.

**Comment by simion1232006** [comment]
> wow! i did not know that WW3 did that. But we do that. that was going to be my next ask. When you create the child grid, in my m file, after you netcdf load the grid file, you can add lon_rho=lon_rho+0.00001; and then continue making the child grid. see if that works, or if that WW3 flag works.

Yes, I would think either one would work. Took me a while to spot this in ww3 manual. 

Confirmed that the flag works.

**Comment by simion1232006** [comment]
> wow! i did not know that WW3 did that. But we do that. that was going to be my next ask. When you create the child grid, in my m file, after you netcdf load the grid file, you can add lon_rho=lon_rho+0.00001; and then continue making the child grid. see if that works, or if that WW3 flag works.


HI John,

Here is my updates:
I can run ww3 stand alone with nesting which has no issues as per plots attached.
I can run ww3 couple with ROMS but the model blow up at the first time step of coupling. I have looked into the rst file which shows odd information along the boundary of ww3 child grid (I reduced the sizse of child grid to have no land masking). I think there could be interpolation issue during coupling. Could you please help?


![image](https://github.com/user-attachments/assets/a8466674-e2ec-4e8a-9706-e5f37b8c0e64)
![image](https://github.com/user-attachments/assets/76cbce6d-04f5-4d86-940c-8225dcc0949f)
![image](https://github.com/user-attachments/assets/04806ebd-87e7-42d8-9fee-dcdd83370668)

**Comment by jcwarner-usgs** [comment]
ok. So i have been working on this and just pushed out some updates for WW3 and the coupling.  I have modified the coupling a bit, and updating the m files.  I have the Sandy case now with ROMS 2 grids and WW3 2 grids coupled and it works well.
Please use roms and ww3 on the same grids (same parent grid and same child grid).  Do not adjust the size of either.
For the create_ww3_grid_files,  we do not mask the ww3 areas based on roms mask.  We also use a boundary of '2' for the parent and child. Also updated the SCRIP routines to make the weights file for roms-ww3.
Can you do an update, make your grids the same size, and try this procedure and see if it works now?
thanks
j

**Comment by simion1232006** [comment]
> ok. So i have been working on this and just pushed out some updates for WW3 and the coupling. I have modified the coupling a bit, and updating the m files. I have the Sandy case now with ROMS 2 grids and WW3 2 grids coupled and it works well. Please use roms and ww3 on the same grids (same parent grid and same child grid). Do not adjust the size of either. For the create_ww3_grid_files, we do not mask the ww3 areas based on roms mask. We also use a boundary of '2' for the parent and child. Also updated the SCRIP routines to make the weights file for roms-ww3. Can you do an update, make your grids the same size, and try this procedure and see if it works now? thanks j

Really Appreciate your effords . I will give a try and let you know how it goes.

**Comment by jcwarner-usgs** [comment]
good.
i just did another update to the SCRIP routines.  had trouble with the unstructured WW3.
so make sure you get a version update now.
If that works, i can walk u thru the unstructured grid for WW3, seems to run faster.

**Comment by simion1232006** [comment]
> good. i just did another update to the SCRIP routines. had trouble with the unstructured WW3. so make sure you get a version update now. If that works, i can walk u thru the unstructured grid for WW3, seems to run faster.

After updating the script interpolation routine, the model is  running all right. I do notice your script  excluded the masking for WW3. I think it is still worth to keep the masking, otherwise waves modelled behind the island would be wrong.
Also I am keen to know how to couple ROMS with the unstructued grid WW3. The model I have uses more time in wave simulation than hydrodynamics. 

![Calypso_L1_2016_Hs002](https://github.com/user-attachments/assets/689a9ebc-5e8a-4cc3-8294-d414e1ccaf20)
![Calypso_L2_2016_Hs002](https://github.com/user-attachments/assets/34e7d437-8ad8-4d10-88e5-3ea66fd07170)

**Comment by jcwarner-usgs** [comment]
That still does not quite look correct.  What are you using for init conditions?  how far into the simulation is that?

For the unstructured, look at Section 11.3.  I wrote that up in the manual.

**Comment by simion1232006** [comment]
> That still does not quite look correct. What are you using for init conditions? how far into the simulation is that?
> 
> For the unstructured, look at Section 11.3. I wrote that up in the manual.

The model is all right. I delibrately initialised the parent grid but not the child grid just to test the nesting. 
see the plot below. 
I am wondering whether it is possible to use the masking or  should I modify the bathymetry to be 0m at land?

![Calypso_L1_2016_Hs012](https://github.com/user-attachments/assets/bf76aaff-1a7a-4691-bf61-9efe4dc31cbd)
![Calypso_L2_2016_Hs012](https://github.com/user-attachments/assets/3a608380-5b00-4012-b46b-24a9dd2bc04a)

**Comment by simion1232006** [comment]
> That still does not quite look correct. What are you using for init conditions? how far into the simulation is that?
> 
> For the unstructured, look at Section 11.3. I wrote that up in the manual.

Hi John, may I ask whether COAWST can do mulitple layers (e.g., three sets of grid) nesting and coupling?

**Comment by jcwarner-usgs** [comment]
yes you can do telescoping nested coupled grids.  I have run up to 4 telescoping grids.
The Sandy can has 2 roms 2 swan(or ww3) and 2 wrf grids.
but you can do more.
typically i would suggest you have the ocn and wav on the same grids.  the atm can have multiple grids and be on different grids.

**Comment by simion1232006** [comment]
> yes you can do telescoping nested coupled grids. I have run up to 4 telescoping grids. The Sandy can has 2 roms 2 swan(or ww3) and 2 wrf grids. but you can do more. typically i would suggest you have the ocn and wav on the same grids. the atm can have multiple grids and be on different grids.

Thanks a lot, John. That is good to know.  Also as a update, I solved the issue of masking. It seems I need to edit the mask of nested grid to ensure consistency. Otherwise the model would not run properly.
